proxy/src/transport/tls/conditional_accept.rs

```diff
@@ -0,0 +1,246 @@
+#![allow(dead_code)] // TODO: Use this.
+
+use futures::Async;
+
+use super::{Identity, untrusted};
+
+/// Determintes whether the given `input` looks like the start of a TLS
+/// connection that the proxy should terminate.
+///
+/// Returns `Async::Ready(true)` if the proxy must process the TLS connection
+/// itself.
+///
+/// Returns `Async::Ready(false)` if the proxy must *not* process the
+/// TLS connection itself.
+///
+/// The determination is made based on whether the input looks like (the start
+/// of) a valid ClientHello that a reasonable TLS client might send, and the
+/// SNI matches the given identity.
+///
+/// XXX: Once the TLS record header is matched, the determination won't be
+/// made until the entire TLS record including the entire ClientHello handshake
+/// message is available. TODO: Reject non-matching inputs earlier.
+///
+/// This assumes that the ClientHello is small and is sent in a single TLS
+/// record, which is what all reasonable implementations do. (If they were not
+/// to, they wouldn't interoperate with picky servers.)
+pub fn match_client_hello(input: &[u8], identity: &Identity) -> Async<bool> {
+    let r = untrusted::Input::from(input).read_all(untrusted::EndOfInput, |input| {
+        let r = extract_sni(input);
+        input.skip_to_end(); // Ignore anything after what we parsed.
+        r
+    });
+    match r {
+        Ok(Some(sni)) => {
+            let matches = if let Ok(sni) = Identity::from_sni_hostname(sni.as_slice_less_safe()) {
+                sni == *identity
+            } else {
+                false
+            };
+            trace!("match_client_hello: parsed correctly up to SNI; matches: {}", matches);
+            Async::Ready(matches)
+        },
+        Ok(None) => {
+            trace!("match_client_hello: failed to parse up to SNI");
+            Async::Ready(false)
+        },
+        Err(untrusted::EndOfInput) => {
+            trace!("match_client_hello: needs more input");
+            Async::NotReady
```

Comment:
While `Async` does model your states nicely, it seems incorrect to use it here:  When I see `Async`, I automatically assume that we are beholden to the constraints of operating within a future: that this must be called from the context of a task, and that the caller will receive a notification when the function shuold be called again.  Granted, it's hard to assume the latter from this API; but I think it would be less confusing to use a one-off enum:


```
pub enum ClientHelloMatch {
  Incomplete,
  Matched,
  NotMatched,
}
```

Or something.

Reply:
OK, I also had similar misgivings. I changed it to be as you suggested, though I shorted the name to `Match`.